processing/src/main/java/org/apache/druid/frame/file/FrameFileHttpResponseHandler.java

```diff
@@ -114,20 +124,34 @@ private ClientResponse<FrameFilePartialFetch> response(
       return ClientResponse.finished(clientResponseObj);
     }
 
-    final byte[] chunk = new byte[content.readableBytes()];
-    content.getBytes(content.readerIndex(), chunk);
+    final byte[] chunk;
+    final int chunkSize = content.readableBytes();
 
-    try {
-      final ListenableFuture<?> backpressureFuture = channel.addChunk(chunk);
+    // Potentially skip some of this chunk, if the relevant bytes have already been read by the handler. This can
+    // happen if a request reads some data, then fails with a retryable I/O error, and then is retried. The retry
+    // will re-read some data that has already been added to the channel, so we need to skip it.
+    final long readByThisHandler = channel.getBytesAdded() - startOffset;
+    final long readByThisRequest = clientResponseObj.getBytesRead(); // Prior to the current chunk
+    final long toSkip = readByThisHandler - readByThisRequest;
 
-      if (backpressureFuture != null) {
-        clientResponseObj.setBackpressureFuture(backpressureFuture);
-      }
+    clientResponseObj.addBytesRead(chunkSize);
```

Comment:
In the skip case we are not basically reading all the bytes. We are skipping to `chunkSize - (int) toSkip]`. So does adding in all the bytes is valid?

Reply:
It is, since "bytes read" tracks how many bytes have been read from the response, not how many have been written to the channel. Its purpose is to let us know when to stop skipping (we keep incrementing it until it passes `channel.getBytesAdded() - startOffset`). I added a comment about this.